src/types.ts

```diff
@@ -4,20 +4,40 @@
  */
 
 /* eslint-disable @typescript-eslint/no-empty-interface */
-
-import type * as express from 'express';
 import type * as http from 'http';
 import type * as httpProxy from 'http-proxy';
 import type * as net from 'net';
 import type * as url from 'url';
 
-export interface Request extends express.Request {}
-export interface Response extends express.Response {}
+export type Request = http.IncomingMessage;
+export type Response = http.ServerResponse;
 
-export interface RequestHandler extends express.RequestHandler {
-  upgrade?: (req: Request, socket: net.Socket, head: any) => void;
+/**
+ * http-proxy-middleware supports framework specific values. The following
+ * values are primarily decorated onto IncomingMessage by express, but are
+ * not required for use.
+ */
+declare module 'http' {
```

Comment:
Augmenting `http.IncomingMessage` does some assumptions on properties from certain frameworks; ie. `express`.

Wondering if there is a way to provide better support (without hardcoding them?)

Reply:
I’m AFK for a day or so, but iirc, HPM mutates and/or attempts to read these values. This augment was only to satisfy HPMs existing internal usage of these fields. It does not require frameworks to implement these, luckily. 

Will have to visit on a field by field basis. Probably worth going thru them and seeing if we can avoid

Reply:
Alright, I had some time this morning.

I was _almost_ able to decouple full from express, and improve typesafety (by reducing casts).

https://github.com/cdaringe/http-proxy-middleware/pull/1/

Unfortunately, there is one macro barrier--and that is that fact that internally `express` _mutates_ `req.url` in a fn called `trim_prefix` and exposes the desired url on `originalUrl`. this is harmful behavior, and make supporting express difficult on a plain node `http`, particularly because apps can `app.use(...)` at many levels inside of `express`.

so, we need to either need to augment and still support `originalUrl`, or, accept a `pathPrefix` (🤢 ) and use that to reconstruct original `req.url`s. The latter sounds dicey, and perhaps has edge cases.

Reply:
I also posted to express here: https://github.com/expressjs/express/issues/4854

Reply:
Did some testing today.

Think augmenting native types is somewhat undesirable; As these augmented types will bubble up to the projects in which http-proxy-middleware is used.

Wouldn't adding back `@types/express` as devDependency makes sense? And cast internal `req` objects to `express.Request` at places where it's needed. Instead creating and maintaining them as internal types. This way it's also documented for which framework it's serving its purpose.

Reply:
> Think augmenting native types is somewhat undesirable; As these augmented types will bubble up to the projects in which http-proxy-middleware is used.

Agreed. Let's not do that.

> Wouldn't adding back @types/express as devDependency makes sense? And cast internal req objects to express.Request at places where it's needed.

Ya, we can do that

Reply:
mmm, in retrospect, if we bring them back and cast, they'd need to be a peer dependency or top level dependency for TS users who aren't `skipLibCheck: true`'ing :).

the only fields that are needed are `body` & `originalUrl`--pretty minimal. to prevent users from having to deal with library and dependency hoopla, i was able to put in _minimal_ types, right around the _runtime_ places where we need to handle `express.js` special cases as well. thus, we can co-locate express compile & runtime special cases simultaneously, adjacent to one another, with no dependency complexity.

let me annotate my new patches to draw attention to this to you--see if you're game

Reply:
Getting into a rabbit-hole... Didn't expect changes/refactoring of the internals. 😬
Need some time to review and see if it's going into the right direction.
It's a tough cookie...

Reply:
Getting into a rabbit-hole... Didn't expect changes/refactoring of the internals. 😬
Need some time to review and see if it's going into the right direction.
It's a tough cookie...

Reply:
> Didn't expect changes/refactoring of the internals


ya... sorry about that. it's much less evil than it looks! mainly just a single param drop, which generated a ton of 🔴 / 🟢  in the diffs. nonetheless, it was all still on topic for de-expressing--promise! 😬 🙏